File: pocketGems/8MaxArray/maxArray.cpp
#include <iostream>
#include <vector>
#include <list>

using namespace std;

struct E
{
    int num;
    int index;
    E(int n, int i)
        : num(n), index(i) {}
};

vector<int> maxArray(vector<int>& nums, int k) 
{
    if (nums.empty())   return vector<int>();
    vector<int> result(nums.size());
    list<E> lt;
    
    for (int i = nums.size() - 1; i >= 0; --i)
    {
        while (!lt.empty() && lt.front().index - i >= k)
            lt.pop_front();
        while (!lt.empty() && lt.back().num <= nums[i])
            lt.pop_back();
        lt.push_back(E(nums[i], i));
        result[i] = lt.front().num;
    }
    return result;
}

int main()
{
    vector<int> v = {1, 3, 5, 7, 3, 4, 2, 9};
    vector<int> result = maxArray(v, 3);
    for (const int &i : result)
        cout << i << " ";
    cout << endl;
}

File: pocketGems/8MaxArray/MaxArray.java
import java.util.*;

class Entry {
	int num, index;
	Entry (int num, int index) {
		this.num = num;
		this.index = index;
	}
}

public class MaxArray {
	public static int[] maxArray(int[] A, int k) {
		Deque<Entry> dq = new LinkedList<>();
		int[] result = new int[A.length];
		for (int i = A.length - 1; i >= 0; i--) {
			while (!dq.isEmpty() && dq.peekFirst().index - k >= i) {
				dq.pollFirst();
			}
			while (!dq.isEmpty() && dq.peekLast().num <= A[i]) {
				dq.pollLast();
			}
			dq.offer(new Entry(A[i], i));
			result[i] = dq.peekFirst().num;
		}
		return result;
	}
	public static void main(String[] args) {
//		int[] A = {1, 3, 5, 7, 3, 4, 2, 9};
		int[] A = {1,2,3,3,6};
		System.out.println(Arrays.toString(maxArray(A, 3)));
	}
}